### ISSUE TITLE
`touch`, `date`: Replace humantime_to_duration with fundu

### ISSUE THREAD

**Issue by Joining7943** [uutils/coreutils#4967]
Like I mentioned in another issue a while ago, `fundu` since version 1.0.0 is capable of parsing gnu relative times as described in the gnu documentation. If you're interested you (or I) could replace the usages of `humantime_to_duration::from_str` in `touch` and `date` with fundu.

**Comment by sylvestre** [comment]
sorry but what would be the gain?

**Comment by sylvestre** [comment]
To be honest, I am more thinking about removing the usage of fundu in coreutils to use our implementation :(

**Comment by Joining7943** [comment]
> sorry but what would be the gain?

fundu is definitely a lot faster than humantime_to_duration and it's well tested. Please also see the [README](https://github.com/Joining7943/fundu). Maybe you could also spare yourself the maintenance of another crate.

**Comment by fritzmark** [comment]
Was here already something done? When I try to update _coreutils_ I get this error:

`Updating coreutils
    Updating crates.io index
  Installing coreutils v0.0.19
    Updating crates.io index
error: failed to compile `coreutils v0.0.19`, intermediate artifacts can be found at `C:\Users\xxxxxxx\AppData\Local\Temp\cargo-installawI5rs`

Caused by:
  no matching package named `humantime_to_duration` found
  location searched: registry `crates-io`
  required by package `uu_date v0.0.19`
      ... which satisfies dependency `date = "^0.0.19"` of package `coreutils v0.0.19`


Updated 0 packages.
Failed to update coreutils.`

**Comment by Joining7943** [comment]
No I haven't done anything yet. I'm waiting on a response. Even after the rename of the crate the gnu relative time parsing to a Duration could be done by `fundu`.

**Comment by sylvestre** [comment]
> Maybe you could also spare yourself the maintenance of another crate.

Well, sorry to be blunt, but given that you are the only contributor to fundu, that it is published on your own github and don't have any other users than the [rust coreutils](https://crates.io/crates/fundu/reverse_dependencies) yet, this isn't appealing to me :(

**Comment by sylvestre** [comment]
> Should this really be done so extreme? 
Because I want to make sure that nobody uses it.

> What happens if one wants to install an older version?
I unyanked the version version needed by 0.0.19

**Comment by tertsdiepraam** [comment]
Hi @Joining7943!

We've discussed among the maintainers and we think we've found a good solution.

Let me first explain why switching to `fundu` is tricky. First, we need to support a very specific format with 100% compatibility, this makes it very likely that we need to tweak the parsing, which might be hard to do if the changes have to go through you. We'll probably also get the bug reports if something is wrong since people might notice the differences with GNU. Therefore, we'd like to have a crate under the uutils organisation, which is `parse_datetime` (previously `humantime_to_duration`). We also do not need the flexibility that `fundu` provides.

Therefore, we propose that we hard fork `fundu` and tailor it to our needs under the uutils organisation. We'll credit you of course and keep your git history. That way, you can keep the `fundu` crate and the flexibility it provides, because it looks very good for many other applications!

We'll probably also use some parsing library, but we'll have to see how that works out. Does that sound good to you?

**Comment by Joining7943** [comment]
@tertsdiepraam Looks like we've commented simultaneously

> Let me first explain why switching to fundu is tricky. We need to support a very specific format with 100% compatibility. This makes it very likely that we need to tweak the parsing, which might be hard to do if the changes have to go through you. We'll probably also get the bug reports if something is wrong since people might notice the differences with GNU.

I actually checked compatibility with gnu (relative time parsing) and haven't found any issues. If there are any, I would be interested, too :) Maybe have a look at the `gnu` example in the examples folder. However, maybe you want to wait for v1.1.0 (it should be ready soon), I'm in the process (currently on the develop branch) to expose the core parser in a separate feature, so you don't have to use the `standard`, `custom` feature and on the other side have influence on the parsing of time units. This'll make the creation of the parser possible at compile time as `const` removing the initialisation time which is not const from the parsing time, makes the parsing of the time units slightly faster and also lowers the final binary size a little bit, if that's a concern. Like I said in the comment before if you have ideas, feature requests, questions, compatibility issues etc. feel free to create an issue on fundu's github page.

> Therefore, we propose that we hard fork fundu and tailor it to our needs under the uutils organisation. We'll credit you of course and keep your git history.

Honestly, I'm not sure which parts you want to tailor to make it more compatible with gnu but feel free to fork it.

**Comment by tertsdiepraam** [comment]
Alright! Thank you! We'll wait for v1.1 and try to fork from there.

**Comment by sylvestre** [comment]
Or would you be ok to move fundu in uutils and maybe we can play with rust features to have it only what we need for it?

**Comment by Joining7943** [comment]
Thank's for your offer, but I would rather keep fundu as it is for now. I created an issue https://github.com/Joining7943/fundu/issues/22 for you. I'm currently restructuring the project anyways and I'll create a package `fundu-core` together with fundu's main package v1.1.0 release. I'm not sure but I guess `fundu-core` is what you wanted, although I still think it's easier and also have 100% compatibility with `gnu` if you would use one of the facades from the `custom` or the (currently in development) `base` feature, which I described earlier in the comment above. Like I said before, feel free to join the discussion at fundu or create an own issue if you have any questions, feature requests etc.

**Comment by sylvestre** [comment]
> Thank's for your offer, but I would rather keep fundu as it is for now. 

Sure I understand why you want this. 
But given you are the only maintainer/contributor, it adds risks for coreutils that I am not sure I want to take currently.
Sorry

**Comment by Joining7943** [comment]
I definitely won't stop anyone from contributing. Although, due to the restructuring of the project it's not the best time to do so. But, as soon as the big restructuring is completed (v1.1.0) it's a great to time to make contributions. Also, if these are just tests to guarantee a specific property. For example, I wrote tests to check against the gnu format but I haven't marked them specifically, so they are included in the rest of the tests ecosystem (which by the way have 100% coverage). Fundu's also tested on critical systems and I'm running fuzzy tests pretty much from the beginning.

**Comment by fritzmark** [comment]
> > Should this really be done so extreme?
> 
> Because I want to make sure that nobody uses it.
> 
> > What happens if one wants to install an older version?
> 
> I unyanked the version version needed by 0.0.19

That action fixed it:

```shell
   Installed package `coreutils v0.0.19` (executable `coreutils.exe`)
```

**Comment by sylvestre** [comment]
well said @piotrkwiecinski

**Comment by TheDcoder** [comment]
@piotrkwiecinski Can't whoever who wants to continue development just fork it in any of those scenario? I don't see the issue as long as the current version is well maintained.

**Comment by simon04** [comment]
Using fundu's `parse_with_date` would fix #4811.

**Comment by tertsdiepraam** [comment]
We have made the decision to expand our own `parse_datetime` crate instead. We would happily accept fixes for #4811 there. Thank you all for your input. I'll close this issue now.
